File: week1_baseline/python/08_the_repl_loop/boukensha/registry.py
from .errors import UnknownToolError
from .tool import Tool


class Registry:
    def __init__(self, context):
        self.context = context

    # Registers a tool on the context. Used as a decorator — the decorated
    # function is the tool's block (Ruby passes it as a `&block` instead).
    def tool(self, name, *, description, parameters=None):
        def register(block):
            self.context.register_tool(
                Tool(str(name), description, parameters or {}, block)
            )
            return block

        return register

    def dispatch(self, name, args=None):
        tool = self.context.tools.get(str(name))
        if not tool:
            raise UnknownToolError(f"No tool registered as '{name}'")
        return tool.block(**(args or {}))

File: week1_baseline/ruby/10_standard_tool_library/lib/boukensha/registry.rb
require_relative "errors"

module Boukensha
  class Registry
    # Exposed so a caller can ask what is already registered before registering.
    # Tools::Mcp needs this to detect name collisions between MCP servers —
    # #tool would otherwise let a second registration silently clobber a first.
    attr_reader :context

    def initialize(context)
      @context = context
    end

    def tool(name, description:, parameters: {}, &block)
      tool = Tool.new(name.to_s, description, parameters, block)
      @context.register_tool(tool)
      tool
    end

    def dispatch(name, args = {})
      tool = @context.tools[name.to_s]
      raise UnknownToolError, "No tool registered as '#{name}'" unless tool
      tool.block.call(**args.transform_keys(&:to_sym))
    end
  end
end